### PR TITLE
Horizontal regions

### PR DESCRIPTION
**Pull request by jdahm**
## Description

Horizontal region GDP implementation in frontend, classic, and GTC toolchains.

To do:
- [ ] Frontend tests
- [ ] Validation tests(?)
- [ ] GTC implementation

Blocked by: #367,  #368

### CODE REVIEW

File: src/gt4py/analysis/passes.py [hunk 1]
```diff
@@ -829,9 +841,76 @@ def apply(transform_data: TransformData) -> None:
...
+                            stmt_infos.append(stmt_info)
+                    int_block.stmts = stmt_infos
+
+
 class ComputeExtentsPass(TransformPass):
     """Loop over blocks backwards and accumulate extents.
 
+    This includes computing extents for the HorizontalIf blocks, and marking these
+    for deletion if they are unused.
+
```
Comment (jdahm) [inline src/gt4py/analysis/passes.py:913]: Remove this.

File: src/gt4py/analysis/passes.py [hunk 2]
```diff
@@ -849,26 +928,38 @@ def apply(transform_data: TransformData) -> None:
...
+                                    list(Extent(extent[:seq_axis]) - diffs) + [(0, 0)]
+                                )
+                                compute_domain_extents[name] |= compute_extent
+
+                            for name in stmt_info.outputs:
+                                compute_domain_extents[name] |= ij_block.compute_extent
 
         transform_data.implementation_ir.fields_extents = {
-            name: Extent(extent) for name, extent in access_extents.items()
+            name: Extent(extent) for name, extent in compute_domain_extents.items()
```
Comment (jdahm) [inline src/gt4py/analysis/passes.py:962]: Rename back to `access_extents`.

File: src/gt4py/backend/python_generator.py [hunk 3]
```diff
@@ -193,25 +193,27 @@ def visit_NativeFuncCall(self, node: gt_ir.NativeFuncCall):
...
+        lhs = self.visit(node.target, **kwargs)
+        rhs = self.visit(node.value, **kwargs)
 
         source = "{lhs} = {rhs}".format(lhs=lhs, rhs=rhs)
 
         # self.var_refs_defined is used in the numpy backend
-        if isinstance(node.target, gt_ir.VarRef):
+        if isinstance(node.target, gt_ir.VarRef) or (
+            hasattr(node.target, "expr") and isinstance(node.target.expr, gt_ir.VarRef)
+        ):
```
Comment (jdahm) [inline src/gt4py/backend/python_generator.py:216]: Revert this change.

File: src/gt4py/ir/nodes.py [hunk 4]
```diff
@@ -713,6 +732,13 @@ def full_interval(cls, order=IterationOrder.PARALLEL):
         return interval
 
 
+# TODO Find a better place for this in the file.
```
Comment (jdahm) [inline src/gt4py/ir/nodes.py:735]: Address TODO.

File: tests/test_unittest/test_compute_extents_pass.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    """
+    overlap = overlap_with_extent(IntervalInfo(start=(1, -3), end=(1, 0)), (0, 1))
+    assert overlap == (-IntervalInfo.MAX_INT, 1)
+
+    overlap = overlap_with_extent(IntervalInfo(start=(1, -1), end=(1, 0)), (0, 0))
+    assert overlap == (-IntervalInfo.MAX_INT, 0)
+
+    overlap = overlap_with_extent(IntervalInfo(start=(0, 2), end=(0, 3)), (0, 0))
+    assert overlap == (-2, IntervalInfo.MAX_INT)
+    """
```
Comment (jdahm) [inline tests/test_unittest/test_compute_extents_pass.py:43]: Convert block comment into tests.

File: src/gt4py/backend/python_generator.py [hunk 6]
```diff
@@ -169,21 +169,21 @@ def visit_VarRef(self, node: gt_ir.VarRef):
...
         rhs_fmt = "({})" if isinstance(node.rhs, gt_ir.CompositeExpr) else "{}"
         source = "{lhs} {op} {rhs}".format(
-            lhs=lhs_fmt.format(self.visit(node.lhs)),
+            lhs=lhs_fmt.format(self.visit(node.lhs, **kwargs)),
             op=self.OP_TO_PYTHON[node.op],
-            rhs=rhs_fmt.format(self.visit(node.rhs)),
+            rhs=rhs_fmt.format(self.visit(node.rhs, **kwargs)),
         )
 
         return source
```
Comment (eddie-c-davis) [inline src/gt4py/backend/python_generator.py:189]: Recommend adding `**kwargs` to `visit_NativeFuncCall`...

Comment (jdahm) [reply src/gt4py/backend/python_generator.py:189]: Good catch!

**Comment by jdahm** [comment]
It seems that this test simply ran out of execution time and ended with `SIGABRT`.
